redash/query_runner/pinot.py

```diff
@@ -0,0 +1,135 @@
+                "controllerURI": {"type": "string", "default": "http://localhost:9000"},
+                "username": {"type": "string"},
+                "password": {"type": "string"},
+            },
+            "order": ["brokerScheme", "brokerHost", "brokerPort", "controllerURI", "username", "password"],
+            "required": ["brokerHost", "controllerURI"],
+            "secret": ["password"],
+        }
+
+    @classmethod
+    def enabled(cls):
+        return enabled
+
+    def __init__(self, configuration):
+        super(Pinot, self).__init__(configuration)
+        self.controller_uri = self.configuration.get("controllerURI")
+        self.username=(self.configuration.get("username") or None)
+        self.password=(self.configuration.get("password") or None)
+
+    def run_query(self, query, user):
+        logger.debug("Running query %s with username: %s, password: %s", query, self.username, self.password)
+        connection = pinotdb.connect(
+            host=self.configuration["brokerHost"],
+            port=self.configuration["brokerPort"],
+            path="/query/sql",
+            scheme=(self.configuration.get("brokerScheme") or "http"),
+            verify_ssl=False,
+            username=self.username,
+            password=self.password,
+        )
+
+        cursor = connection.cursor()
+
+        try:
+            cursor.execute(query)
+            logger.debug("cursor.schema = %s",cursor.schema)
+            columns = self.fetch_columns(
+                [(i["name"], PINOT_TYPES_MAPPING.get(i["type"], None)) for i in cursor.schema]
+            )
+            rows = [
+                dict(zip((column["name"] for column in columns), row)) for row in cursor
+            ]
+
+            data = {"columns": columns, "rows": rows}
+            error = None
+            json_data = json_dumps(data)
+            logger.debug("Pinot execute query [%s], and resp: %s", query, json_data)
+        finally:
+            connection.close()
+
+        return json_data, error
+
+    def get_schema(self, get_stats=False):
+        schema = {}
+        for schema_name in self.get_schema_names():
+            for table_name in self.get_table_names():
+                schema_table_name = "{}.{}".format(schema_name, table_name)
+                if table_name not in schema:
+                    schema[schema_table_name] = {"name": schema_table_name, "columns": []}
+                table_schema =self.get_pinot_table_schema(table_name)
+
+                for column in table_schema.get("dimensionFieldSpecs", []) + table_schema.get(
+                    "metricFieldSpecs", []) + table_schema.get("dateTimeFieldSpecs", []):
+                    c = {
+                            "name": column["name"],
+                            "type": PINOT_TYPES_MAPPING[column["dataType"]],
+                        }
+                    schema[schema_table_name]["columns"].append(c)
```

Comment:
Nit: is there a way to pull the entire schema in one roundtrip to Pinot? This approach of looping through schemas -> tables -> columns will _work_ but scales linearly as a schema grows. This isn't a blocker to merge but perhaps a way to incrementally improve in the future.

Looking at the Pinot docs it seems we could pull the entire schema in one roundtrip: https://docs.pinot.apache.org/users/tutorials/schema-evolution#get-the-existing-schema

Reply:
Pinot doesn't support fetch all schemas at once, so this is the only way.

<img width="1456" alt="image" src="https://user-images.githubusercontent.com/1202120/188975962-a514d775-03c1-46df-85d5-0ba535ecfafe.png">